### PR TITLE
Blog - Build a Data Dashboard with Streamlit in Python

### PR DESCRIPTION
**Pull request by msali123**
Done with all checks.

### CODE REVIEW

**Comment by msali123** [comment]
Preview URL:
https://63cb83e7bf022d307bcf2310--ephemeral-narwhal-3ac3b2.netlify.app

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+~~~
+
+To check if the installation worked, create a new Python script in your editor, import streamlit under the alias `st`, and use the `write` function to print out some text. The `st.write()` function is used to display information like text, dataframes, or figures.
+
+~~~{.bash caption=">_"}
+import streamlit as st
+st.write("Hello World!")
+st.write("Hello Streamlit!")
+
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:47]:
```suggestion

~~~{.python caption="data_analysis.py"}
import streamlit as st
st.write("Hello World!")
st.write("Hello Streamlit!")

~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+`st.pyplot()` is the Streamlit function to create figure objects and plots. To create a [matplotlib](/blog/plotting-rainfall-data-with-python-and-matplotlib) visualization, you have to perform data analysis and then create the visualization. For this section, we will use [this movie industry dataset from
+Github](https://github.com/danielgrijalva/movie-stats).
+
+This dataset contains over 7000 movie entries—from the period 1986-2016—scraped from IMDb (Internet Movie Database). It lists movies of different genres and countries. I'll be using this dataset to create different interactive plots for this tutorial.
+
+~~~{.python caption=""}
+import streamlit as st
+import pandas as pd
+import matplotlib.pyplot as plt
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:79]:
```suggestion

~~~{.python caption="data_analysis.py"}
import streamlit as st
import pandas as pd
import matplotlib.pyplot as plt
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+We use the pandas `read_csv()` function to read in the data into a dataframe.
+
+~~~{.python caption=""}
+#read in the file
+movies_data = pd.read_csv("https://raw.githubusercontent.com/danielgrijalva/movie-stats/7c6a562377ab5c91bb80c405be50a0494ae8e582/movies.csv")
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:86]:
```suggestion

~~~{.python caption="data_analysis.py"}
#read in the file
movies_data = pd.read_csv("https://raw.githubusercontent.com/danielgrijalva/movie-stats/7c6a562377ab5c91bb80c405be50a0494ae8e582/movies.csv")
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md (hunk 1 above)
Comment (balapriyac) [reply blog/_posts/2023-02-19-streamlit-python-dashboard.md:47]: The author submitted a GitHub link to their code for this tutorial. All code is in the data_analysis.py file, so adding it to the caption.

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+To generate a summary of the dataset and check for missing values and duplicates, we'll use the following functions:
+
+~~~{.python caption=""}
+movies_data.info()
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:92]:
```suggestion

~~~{.python caption="data_analysis.py"}
movies_data.info()
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+
+~~~{.python caption="Output"}
+#output
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:95]:
```suggestion
~~~{caption="Output"}
#output
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+ 13  company   7651 non-null   object 
+ 14  runtime   7664 non-null   float64
+dtypes: float64(5), int64(1), object(9)
+~~~
+
+As seen, `movies_data.info()` gives a quick overview of our dataset. We can see that there are 7668 entries (rows) and a total of 15 columns.
+
+~~~{.python caption=""}
+movies_data.duplicated()
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:123]:
```suggestion

~~~{.python caption="data_analysis.py"}
movies_data.duplicated()
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+
+~~~{.python caption="Output"}
+#output
+0       False
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:127]:
```suggestion

~~~{caption="Output"}
#output
0       False
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 8]
```diff
@@ -0,0 +1,355 @@
...
+7666    False
+7667    False
+Length: 7668, dtype: bool
+~~~
+
+The method `movies_data.duplicated()` checks if there are any duplicates. All rows returned `False` which means there are no duplicates.
+
+~~~{.python caption=""}
+movies_data.count()
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:145]:
```suggestion

~~~{.python caption="data_analysis.py"}
movies_data.count()
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+
+~~~{.python caption="Output"}
+#output
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:148]:
```suggestion

~~~{ caption="Output"}
#output
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+gross       7479
+company     7651
+runtime     7664
+~~~
+
+Calling the `count()` method on the dataframe: `movies_data.count()` returns the sum of all entries in a column. Columns with less than 7668 entries suggest missing values.
+
+~~~{.python caption=""}
+movies_data.dropna()
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:170]:
```suggestion

~~~{.python caption="data_analysis.py"}
movies_data.dropna()
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+
+~~~{.python caption=""}
+st.write("""
+Average Movie Budget, Grouped by Genre
+""")
+avg_budget = movies_data.groupby('genre')['budget'].mean().round()
+avg_budget = avg_budget.reset_index()
+genre = avg_budget['genre']
+avg_bud = avg_budget['budget']
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:182]:
```suggestion

~~~{.python caption="data_analysis.py"}
st.write("""
Average Movie Budget, Grouped by Genre
""")
avg_budget = movies_data.groupby('genre')['budget'].mean().round()
avg_budget = avg_budget.reset_index()
genre = avg_budget['genre']
avg_bud = avg_budget['budget']
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 12]
```diff
@@ -0,0 +1,355 @@
...
+
+The `groupby` method groups data by categories using the columns of a dataset and applies a function to it. Here we group by the 'genre' and the 'budget'. And we apply the  `mean()` and the `round()` functions. The `mean()` function returns the average of a list of numbers while the `round()` function rounds up digits and returns a float. The `reset_index()` method resets the index of an updated dataframe; creating a new row index that starts at 0. Resetting indexes is important so [pandas](/blog/plotting-rainfall-data-with-python-and-matplotlib) can find the indexes of elements.
+
+~~~{.python caption=""}
+fig = plt.figure(figsize = (19, 10))
+
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:188]:
```suggestion

~~~{.python caption="data_analysis.py"}
fig = plt.figure(figsize = (19, 10))

```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+plt.ylabel('budget')
+plt.title('Matplotlib Bar Chart Showing the Average \
+Budget of Movies in Each Genre')
+~~~
+
+[Matplotlib](/blog/plotting-rainfall-data-with-python-and-matplotlib) has a function called `show()` that creates a figure object. In Streamlit, this line should be replaced with `st.pyplot(variable_name)` where `variable_name` is the variable of visualization.
+
+~~~{.python caption=""}
+st.pyplot(fig)
+
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:200]:
```suggestion

~~~{.python caption="data_analysis.py"}
st.pyplot(fig)

```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 14]
```diff
@@ -0,0 +1,355 @@
...
+Columns in Streamlit operate just as they do in documents and on web pages. They are also highly responsive and automatically resize on different screens.
+
+To create columns, simply assign them to the variables that match the number of columns you need. Here, `col1` and   `col2` are the variable names because we need two columns.
+
+~~~{.python caption=""}
+col1, col2 = st.columns(2)
+col1.write('# This is Column 1')
+col2.write('# This is Column 2')
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:223]:
```suggestion

~~~{.python caption="data_analysis.py"}
col1, col2 = st.columns(2)
col1.write('# This is Column 1')
col2.write('# This is Column 2')
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 15]
```diff
@@ -0,0 +1,355 @@
...
+<div class="wide">
+![Streamlit Even Columns]({{site.images}}{{page.slug}}/columns_of_same_size.png)\
+</div>
+
+We can as well create columns of different dimensions, where columns are of different sizes.  
+
+~~~{.python caption=""}
+st.write('### Columns of different sizes')
+col1, col2, col3, col4 = st.columns([1,3,1,2])
+
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:234]:
```suggestion

~~~{.python caption="data_analysis.py"}
st.write('### Columns of different sizes')
col1, col2, col3, col4 = st.columns([1,3,1,2])

```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+To link data to a widget, we first convert the needed column to a unique list. This is important so only unique values are selected:
+
+~~~{.python caption=""}
+# Creating sidebar widget unique values from our movies dataset
+score_rating = movies_data['score'].unique().tolist()
+genre_list = movies_data['genre'].unique().tolist()
+year_list = movies_data['year'].unique().tolist()
+
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:273]:
```suggestion

~~~{.python caption="data_analysis.py"}
# Creating sidebar widget unique values from our movies dataset
score_rating = movies_data['score'].unique().tolist()
genre_list = movies_data['genre'].unique().tolist()
year_list = movies_data['year'].unique().tolist()

~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+
+### Creating a Sidebar and Adding Other Widgets
+
+#### Create Layouts Using `with` Statements
+
+The `with` statement provides a simpler, more organized way of displaying Streamlit layouts especially if multiple widgets or variables are attached to a layout component. Using `with` makes the code easier to maintain.
+
+We use the `with` statement to group all elements of a layout together. We've implemented it with a sidebar layout, as shown:
+
+~~~{.python caption=""}
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:283]:
```suggestion

~~~{.python caption="data_analysis.py"}
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+A selectbox is created by calling the `st.selectbox()` function. The selectbox represents a drop-down menu that allows *only one* option to be picked at a time. We linked the 'year' column to this widget, so we can only pick one year at a time.
+
+![Sidebar Image]({{site.images}}{{page.slug}}/sidebar2.png)
+
+To add interactivity among the slider, the selectbox, the multiselect widgets, and the plots on the main page, we need to create filters. We do this by mapping the columns of the dataframe to their unique list and using it in the analysis. By doing so, we can ensure that only the selected widgets affect a plot.
+
+~~~{.python caption=""}
+#Configure and filter the slider widget for interactivity
+score_info = (movies_data['score'].between(*new_score_rating))
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:315]:
```suggestion

~~~{.python caption="data_analysis.py"}
#Configure and filter the slider widget for interactivity
score_info = (movies_data['score'].between(*new_score_rating))
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+
+We will be linking the slider widget to the line chart that displays the number of movies in a particular genre that have scores that fall within a specified range. We mapped the 'score' column to the slider widget. Therefore, whenever a user interacts with the slider, the line chart changes as well.
+
+~~~{.python caption=""}
+#Filter the selectbox and multiselect widget for interactivity
+new_genre_year = (movies_data['genre'].isin(new_genre_list)) \
+& (movies_data['year'] == year)
+~~~
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:323]:
```suggestion

~~~{.python caption="data_analysis.py"}
#Filter the selectbox and multiselect widget for interactivity
new_genre_year = (movies_data['genre'].isin(new_genre_list)) \
& (movies_data['year'] == year)
~~~
```

File: blog/_posts/2023-02-19-streamlit-python-dashboard.md [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+
+We need the multiselect widget and the selectbox that holds genre and year to work together. We will be creating a dataframe that changes movie titles according to the year and genre(s) selected. In our configuration, we mapped the 'genre' column to the variable of our multiselect widget and mapped the 'year' column to the variable of our selectbox widget and joined them both using 'and'.
+
+~~~{.python caption=""}
+# visualization section
```
Comment (balapriyac) [inline blog/_posts/2023-02-19-streamlit-python-dashboard.md:328]:
```suggestion

~~~{.python caption="data_analysis.py"}
# visualization section
```

**Review by balapriyac — CHANGES_REQUESTED** [review]
@jalletto I've finished reviewing this PR.

**Comment by msali123** [comment]
@balapriyac  Requested changes done.

**Pull request merged**